### DISCUSSION TITLE
[Vue] Allow inject in callbacks by implementing runWithContext

### DISCUSSION THREAD

**Discussion by JoostKersjes** [TanStack/query#5424] [Ideas]
I was recently playing around with [provide/inject](https://vuejs.org/guide/components/provide-inject.html) and discovered that inject does not work in callbacks like query `onSuccess`.

Vue 3.3 introduced `app.runWithContext()` (https://github.com/vuejs/core/pull/7451) which allows inject to work outside of components.

`vue-router` implemented `runWithContext` so their global route guards can use inject: https://github.com/vuejs/router/commit/909c45be199fbf48e55f3f7e681e93c259e82a14

**Comment by DamianOsipiuk** [comment]
Callback on `useQuery` will be removed in v5 as they are not reliable.
You can read more about it in this blog post https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose#a-bad-api

**Comment by marlokessler** [comment]
Hi @DamianOsipiuk, what do you recommend as alternative when combining `useQuery` with `provide`?

**Reply by DamianOsipiuk** [reply]
Not sure if i understand the question.

If you are asking about running in `runWithContext`, it is supported from https://github.com/TanStack/query/releases/tag/v4.34.0 and v5.

You can also provide `queryClient` manually as a second parameter to `useQuery`, but you have to be careful about memory leaks, so make sure to wrap usage with [custom effect scope](https://vuejs.org/api/reactivity-advanced.html#effectscope).
